internal/testrunner/coverage_output.go

```diff
@@ -0,0 +1,248 @@
+}
+
+type coberturaLines []*coberturaLine
+
+func (c *coberturaCoverage) bytes() ([]byte, error) {
+	out, err := xml.MarshalIndent(&c, "", "  ")
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to format test results as xUnit")
+	}
+
+	var buffer bytes.Buffer
+	buffer.WriteString(xml.Header)
+	buffer.WriteString("\n")
+	buffer.WriteString(coverageDtd)
+	buffer.WriteString("\n")
+	buffer.Write(out)
+	return buffer.Bytes(), nil
+}
+
+// WriteCoverage function calculates test coverage for the given package.
+// It requires to execute tests for all data streams (same test type), so the coverage can be calculated properly.
+func WriteCoverage(packageRootPath, packageName string, testType TestType, results []TestResult) error {
+	details, err := collectTestCoverageDetails(packageRootPath, packageName, testType, results)
+	if err != nil {
+		return errors.Wrap(err, "can't collect test coverage details")
+	}
+
+	report := transformToCoberturaReport(details)
+
+	err = writeCoverageReportFile(report, packageName)
+	if err != nil {
+		return errors.Wrap(err, "can't write test coverage report file")
+	}
+	return nil
+}
+
+func collectTestCoverageDetails(packageRootPath, packageName string, testType TestType, results []TestResult) (*testCoverageDetails, error) {
+	withoutTests, err := findDataStreamsWithoutTests(packageRootPath, testType)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't find data streams without tests")
+	}
+
+	details := newTestCoverageDetails(packageName, testType).
+		withUncoveredDataStreams(withoutTests).
+		withTestResults(results)
+	return details, nil
+}
+
+func findDataStreamsWithoutTests(packageRootPath string, testType TestType) ([]string, error) {
+	dataStreamDir := filepath.Join(packageRootPath, "data_stream")
+	dataStreams, err := ioutil.ReadDir(dataStreamDir)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't list data streams directory")
+	}
+
+	var noTests []string
+	for _, dataStream := range dataStreams {
+		if !dataStream.IsDir() {
+			continue
+		}
+
+		dataStreamTestPath := filepath.Join(packageRootPath, "data_stream", dataStream.Name(), "_dev", "test", string(testType))
+		_, err := os.Stat(dataStreamTestPath)
+		if errors.Is(err, os.ErrNotExist) {
+			noTests = append(noTests, dataStream.Name())
+			continue
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "can't stat path: %s", dataStreamTestPath)
+		}
+	}
+	return noTests, nil
+}
+
+func transformToCoberturaReport(details *testCoverageDetails) *coberturaCoverage {
+	var classes []*coberturaClass
+	for dataStream, testCases := range details.dataStreams {
+		var methods []*coberturaMethod
+
+		if len(testCases) == 0 {
+			methods = append(methods, &coberturaMethod{
+				Name:  "no-test",
+				Lines: []*coberturaLine{{Number: 1, Hits: 0}},
+			})
+		} else {
+			for i, tc := range testCases {
+				methods = append(methods, &coberturaMethod{
+					Name:  tc,
+					Lines: []*coberturaLine{{Number: i + 1, Hits: 1}},
```

Comment:
Thoughts for a possible future improvement. Coverage is at the data stream level now, right? Maybe in the future we can do it at least at the file level, e.g. saying that pipeline tests add coverage for pipelines, but not for the manifest or other config files, and system tests add coverage for all the files.

Reply:
Ah I know what you mean... We shouldn't mark a missing pipeline test if there are no pipelines at all. Sounds like a follow up.

Reply:
Yes, this would be for the future for sure, no need to do anything else by now :slightly_smiling_face:

Reply:
https://github.com/elastic/elastic-package/issues/442